### PR TITLE
fix f8af582: avoid shared build rule (only share recipe)

### LINKED ISSUE #36: Docker containers install only one python module
Pretty bizarre behavior, we have an error in our setup somewhere. Apparently only one python module per docker image is installed. This only concerns python modules, projects like im6convert, olena and tesseract are installed fine.

In `minimum`, only `ocrd_tesserocr` is installed.

In `medium`, only `cor-asv-ann` is installed.

In `maximum`, only `ocrd_kraken` (of all modules) is installed.

### PR DESCRIPTION
**Pull request by bertsky**
Fixes #36

### CODE REVIEW

**Comment by kba** [comment]
I still don't get the full gamut of modules for `-maximum`:

```sh
docker run --rm -it -- ocrd/all:maximum bash -c 'find / |grep ocrd-'
/usr/local/bin/ocrd-im6convert
/usr/local/bin/ocrd-olena-binarize
/usr/local/share/ocrd_olena/ocrd-tool.json
/usr/local/share/ocrd-im6convert
/usr/local/share/ocrd-im6convert/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd_repair_inconsistencies/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd_ocropy/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd_pc_segmentation/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd_calamari/ocrd-tool.json
/usr/lib/python3.6/site-packages/qurator/sbb_textline_detector/ocrd-tool.json
/usr/lib/python3.6/site-packages/qurator/dinglehopper/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/METADATA
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/WHEEL
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/INSTALLER
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/entry_points.txt
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/RECORD
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/top_level.txt
/usr/bin/ocrd-ocropy-segment
/usr/bin/ocrd-calamari-recognize
/usr/bin/ocrd-dinglehopper
/usr/bin/ocrd-pc-segmentation
/usr/bin/ocrd-sbb-textline-detector
/usr/bin/ocrd-repair-inconsistencies
```

E.g. ocrd_cis is not being included. Could it be that this ONLY contains the additions to `-medium`?

**Comment by kba** [comment]
Tried `mimimum`:

```zsh
$ make docker-minimum
$ docker run --rm -it -- ocrd/all:minimum bash -c 'find / |grep ocrd-'
/usr/local/bin/ocrd-im6convert
/usr/local/share/ocrd-im6convert
/usr/local/share/ocrd-im6convert/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd_repair_inconsistencies/ocrd-tool.json
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/METADATA
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/WHEEL
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/INSTALLER
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/entry_points.txt
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/RECORD
/usr/lib/python3.6/site-packages/ocrd-2.1.2.dist-info/top_level.txt
/usr/bin/ocrd-repair-inconsistencies
```

**Comment by bertsky** [comment]
Sorry, I just saw I also had to move the definition of `multirule` upwards before its first use.

**Comment by stweil** [comment]
That seems to fix it - at least `make all` is now building the missing executables.

**Comment by stweil** [comment]
I still have an (unrelated) build problem with cor-asv-fst (see #37), but the rest seems to work again:

    src/pywrapfst.cc:6610:42: error: ‘class fst::script::WeightClass’ has no member named ‘Member’
    [...]

**Comment by bertsky** [comment]
Regarding the failure with cor-asv-fst (which I'll probably soon delete anyway), there are multiple problems with the installation requirements, including a missing deps-ubuntu. Should I add to this PR, or make a new one?

**Comment by kba** [comment]
> Should I add to this PR, or make a new one?

I want to merge this as soon as my build completes (I'm confident), so please a new PR, thx!

**Comment by bertsky** [comment]
> I want to merge this as soon as my build completes (I'm confident)

The problem is: docker-maximum won't complete, because cor-asv-fst does not install (anymore).

Its installation was buggy before (it would not work like that on a vanilla system), but you would not have noticed until you ran the processor. Now that `pynini` (from OpenFST) has reclaimed the name on PyPI (so we at least pull the correct package!), its full dependencies are exposed: it needs libfst-dev, and a more modern one than Ubuntu 18.04 has.

**Comment by kba** [comment]
> I want to merge this as soon as my build completes (I'm confident), so please a new PR, thx!

Oh, that will prevent the builds actually. In this case, please amend the PR.

**Comment by kba** [comment]
> its full dependencies are exposed: it needs libfst-dev, and a more modern one than Ubuntu 18.04 has.

Remove it, at least temporarily?

**Comment by bertsky** [comment]
> Remove it, at least temporarily?

Just for docker-maximum, or completely from ocrd_all?

You or me?

**Comment by stweil** [comment]
I suggest to filter it out of `OCRD_MODULES`. Please merge PR #39 which does that for opencv-python first.

**Comment by bertsky** [comment]
> I suggest to filter it out of `OCRD_MODULES`. Please merge PR #39 which does that for opencv-python first.

That's a third, mid-way option. If we pre-filter the default `OCRD_MODULES`, we should document this, though. (The expectation and documentation still is that you can use that variable to reduce the modules, not also to increase it.)

**Comment by bertsky** [comment]
Hm, now we have a mix of both methods:
- opencv-python gets removed from the default `OCRD_MODULES` (and therefore also all docker configs)
- cor-asv-fst is removed from docker-maximum(-git) only (and therefore still attempted to build on a local checkout)

**Comment by kba** [comment]
> opencv-python gets removed from the default OCRD_MODULES (and therefore also all docker configs)

But aren't the docker configs (`DOCKER_MODULES`) set explicitly? Except for maximum.

I've not set this up to filter `cor-asv-fst` and `opencv-python` from both the default `OCRD_MODULES` and the `DOCKER_MODULES` for `docker-maximum{,-git}`. This should prevent cor-asv-fst from being included in either `make all` or `make docker-maximum{,-git}`. IIUC.

File: Makefile [hunk 1]
```diff
@@ -31,13 +31,15 @@ PKG_CONFIG_PATH := $(VIRTUAL_ENV)/lib/pkgconfig
...
 CUSTOM_DEPS = wget python3-venv # add more packages for deps-ubuntu below (or modules as preqrequisites)
 
 # Default to all submodules, but allow overriding by user
 # (and treat the empty value as if it was unset)
+# opencv-python is only needed for aarch64-linux-gnu and other less common platforms,
+# so don't include it by default.
+# XXX Wed Jan  8 13:21:44 CET 2020 kba - disabled cor-asv-fst since it won't build on 18.04 without hacks
 ifeq ($(strip $(OCRD_MODULES)),)
-override OCRD_MODULES := $(shell git submodule status | while read commit dir ref; do echo $$dir; done)
+override OCRD_MODULES := $(filter-out cor-asv-fst,$(filter-out opencv-python,$(shell git submodule status | while read commit dir ref; do echo $$dir; done)))
```
Comment (bertsky) [inline Makefile]:
```suggestion
override OCRD_MODULES := $(filter-out cor-asv-fst opencv-python,$(shell git submodule status | while read commit dir ref; do echo $$dir; done))
```

File: Makefile [hunk 2]
```diff
@@ -453,7 +467,8 @@ docker-minimum docker-minimum-git: DOCKER_MODULES = core ocrd_im6convert ocrd_ci
 # Medium-size selection: add Olena binarization and Calamari, use Tesseract from git, add evaluation
 docker-medium docker-medium-git: DOCKER_MODULES = core ocrd_im6convert format-converters ocrd_cis ocrd_tesserocr tesserocr tesseract ocrd_olena ocrd_segment ocrd_keraslm ocrd_calamari dinglehopper cor-asv-ann workflow-configuration ocrd_repair_inconsistencies
 # Maximum-size selection: use all modules
-docker-maximum docker-maximum-git: DOCKER_MODULES = $(filter-out opencv-python,$(OCRD_MODULES))
+# XXX Wed Jan  8 13:21:44 CET 2020 kba - disabled cor-asv-fst since it won't build on 18.04 without hacks
+docker-maximum docker-maximum-git: DOCKER_MODULES = $(filter-out cor-asv-fst,$(filter-out opencv-python,$(OCRD_MODULES)))
```
Comment (bertsky) [inline Makefile]: This is redundant. If the excluded modules are already filtered in the default value for `OCRD_MODULES`, then there's no need to exclude them once more here – except if you want to forbid users to include them again in their own choice of `OCRD_MODULES`.

```suggestion
docker-maximum docker-maximum-git: DOCKER_MODULES = $(OCRD_MODULES)
```

**Comment by bertsky** [comment]
> But aren't the docker configs (`DOCKER_MODULES`) set explicitly? Except for maximum.

Where's the contradiction? Yes, docker-minimum/medium DOCKER_MODULES are set explicitly (without cor-asv-fst or opencv-python anyway), and docker-maximum from OCRD_MODULES. So whatever you exclude in the latter is already excluded in the former.

Plus you don't want to forbid users to enable these in their `OCRD_MODULES` – otherwise you might as well remove them from the repo entirely.

File: Makefile (hunk 1 above)
Comment (kba) [reply Makefile]: interesting, learned something new. I still cannot accept suggestions, so feel free to accept yourself.

File: Makefile (hunk 2 above)
Comment (kba) [reply Makefile]: > except if you want to forbid users to include them again in their own choice of OCRD_MODULES

Yeah, that was the idea. But now that you mention it, it's pretty unlikely. Again, please commit suggestion yourself as I cannot.

**Pull request merged**